---
title: "short_coding_exercise"
author: "Akhil Kumar"
date: "`r Sys.Date()`"
output: 
  html_document:
    code_folding: hide
    keep_md: true
    toc: true
    number_sections: true
    toc_depth: 3
    toc_float: true
    theme: readable
    highlight: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Libraries

## Installations

```{r}
# if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
#
# BiocManager::install("GEOquery")

# install.packages(c("R.utils", "Seurat", "dplyr", "patchwork"))

```

## Loading

```{r warning=FALSE, message=FALSE}
library(GEOquery)

library(R.utils)

library(Seurat)
library(dplyr)
library(patchwork)
# library(scCustomize)
```

# Data retrieval

```{r}
setwd("/Users/iamakhilverma/Desktop/short_exercise_tsankov_lab")
getwd()
```


GEO accession: `GSE125449`
A total of 19 tumors were profiled. Set 1 contains scRNA-seq data of twelve samples, i.e., S16_P10_LCP18, S02_P01_LCP21, S10_P05_LCP23, S09_P04_LCP25, S08_P03_LCP26, S07_P02_LCP28, S11_P06_LCP29, S12_P07_LCP30, S20_P12_LCP35, S21_P13_LCP37, S15_P09_LCP38, and S19_P11_LCP39. Set 2 includes scRNA-seq data of seven samples, i.e., S351_P10_LCP34, S355_P13_LCP42, S358_P16_LCP46, S305_P06_LCP56, S300_P02_LCP60, 364_P21_LCP65, and S365_P22_LCP66. Detailed information can be found in samples.txt file of each Set.

`GPL18573`	Illumina NextSeq 500 (Homo sapiens)
`GPL20301`	Illumina HiSeq 4000 (Homo sapiens)

We care about the data contained in Set 1, i.e., `GSE125449-GPL18573_series_matrix.txt.gz`.

```{r}
# gse <- getGEO('GSE125449',GSEMatrix=TRUE)
# show(gse)
```
```{r}
options(timeout = max(300, getOption("timeout")))
# options(download.file.method.GEOquery = "wget")
```

```{r}
if (!(
  file.exists(
    "/Users/iamakhilverma/Desktop/short_exercise_tsankov_lab/GSE125449/"
  )
)) {
  filePaths <- getGEOSuppFiles('GSE125449')
  filePaths
}
```
```{r}
unzipper <- function(filename) {
  gunzip(
    filename,
    destname = paste(
      '/Users/iamakhilverma/Desktop/short_exercise_tsankov_lab/GSE125449/gse125449_set1/',
      gsub("^GSE125449_Set1_", "", tail(strsplit(
        gsub("[.]gz$", "", filename), '/'
      )[[1]], n = 1)),
      sep = ''
    ),
    remove = FALSE
  )
}
```


```{r}
if (!(
  file.exists(
    "/Users/iamakhilverma/Desktop/short_exercise_tsankov_lab/GSE125449/gse125449_set1"
  )
)) {
  filenames <-
    list.files(path = "/Users/iamakhilverma/Desktop/short_exercise_tsankov_lab/GSE125449/",
               pattern = "GSE125449_Set1_[A-Za-z0-9.]+.gz",
               full.names = TRUE)
  filenames
  sapply(filenames, unzipper)
}
```

The MTX format is a sparse matrix format with genes on the rows and cells on the columns as output by Cell Ranger. You should also have TSV files with genes and barcode sequences corresponding to row and column indices, respectively. These files must be located in the same folder as the MTX file with the same base file name. For example if the MTX file is my_data.mtx, you should also have a my_data.genes.txt file and a my_data.barcodes.txt file.


```{r}
# Load the liver cancer set 1 dataset
livcan.data <- Read10X(data.dir = "./GSE125449/gse125449_set1/")
dim(livcan.data)
# Genes were retained with detected expression in more than 0.1% of all the cells (Total cells in our matrix was 5115, hence 0.1% of 5115 is 511.5). Cells with less than 500 detected genes were excluded.
livcan <-
  CreateSeuratObject(
    counts = livcan.data,
    # min.cells = 511,
    min.cells = 6,
    min.features = 500,
    project = "livcan_set1",
    names.delim = "-",
    names.field = 2
  )
dim(livcan)
# head(livcan.data)
```
"A total of 5,115 cells passed the initial quality control."
Indeed.

# Preprocessing

## QC and selecting cells for further analysis

"We also performed additional quality control steps including choosing the minimum UMI counts (> 700), removing doublets (0.4%, according to the multiplet rate provided by 10x Genomics based on the number of cells loaded and recovered), filtering cells with the percentage of mitochondrial genes (> 20%) (Vento-Tormo et al., 2018)."

Mitochondrial genes can be used to identify low-quality cells. In cells which are broken, cytoplasmic RNA is diluted and RNA which is enclosed in the mitochondria will be retained (Ilicic T. et al. 2016 Classification of low quality cells from single-cell RNA-seq data. Genome Biol.).
Mitochondrial genes starts with prefix “MT-”.

NOTE: The pattern provided (“^MT-“) works for human gene names. You may need to adjust depending on your organism of interest. If you weren’t using gene names as the gene ID, then this function wouldn’t work. 

```{r}
# Show QC metrics for the first 5 cells
head(livcan@meta.data, 5)
livcan[["percent.mt"]] <-
  PercentageFeatureSet(livcan, pattern = "^MT-")
head(livcan@meta.data, 5)


# Visualize QC metrics as a violin plot
VlnPlot(
  livcan,
  features = c("nFeature_RNA", "nCount_RNA", "percent.mt"),
  ncol = 3
)
```

- nFeature_RNA is the number of genes detected in each cell
- nCount_RNA is the total number of molecules (UMIs) detected within a cell

Low nFeature_RNA for a cell indicates that it may be dead/dying or an empty droplet. High nCount_RNA and/or nFeature_RNA indicates that the "cell" may in fact be a doublet (or multiplet).

```{r}
dim(livcan)
# livcan <- subset(livcan, subset = nCount_RNA > 700 & percent.mt < 20)
livcan <- subset(livcan,
                 subset = (nCount_RNA > 700) &
                   # (nFeature_RNA < 2500) &
                   (percent.mt < 20))
dim(livcan)
```

"We found that additional filtering steps did not change an overall cellular composition among these samples."
In our case, No. of cells passing the quality control dropped from 5115 to 4321!

I added the nFeature_RNA cutoffs to take care of any empty droplets or doublets or multiplets. Also, instead of the standard `nFeature_RNA > 200 & nFeature_RNA < 2500`, I used only `nFeature_RNA < 2500` as we only have cells with at least 500 genes in them because of a previous filtration step.
Even if I wouldn't have added the nFeature_RNA cutoffs, the no. of cells passing the quality control was ~4700.

More on doublet detection and removal can be found on- 
1. http://bioconductor.org/books/3.16/OSCA.advanced/doublet-detection.html
2. https://nbisweden.github.io/workshop-scRNAseq/labs/compiled/seurat/seurat_01_qc.html#Predict_doublets


## Normalizing the data

"The total number of transcripts in each single-cell was normalized to 10,000, followed by log transformation, i.e., NormalizedExp = log (RawExp/TotalExp * 10000+1), where RawExp, NormalizedExp and TotalExp represent the raw gene expression value, normalized gene expression as well as the total expression of all genes in a single-cell, respectively."

This step is just using the default scaling factor of 10,000 in the "LogNormalize" global-scaling normalization method.

```{r}
dim(livcan)
livcan <-
  NormalizeData(livcan,
                normalization.method = "LogNormalize",
                scale.factor = 10000)
# # Same can be achieved with the following as we're just using the default values
# livcan <- NormalizeData(livcan)
dim(livcan)
```

## Identification of highly variable features (feature selection)

We next calculate a subset of features that exhibit high cell-to-cell variation in the dataset (i.e, they are highly expressed in some cells, and lowly expressed in others). Researchers have found that focusing on these genes in downstream analysis helps to highlight biological signal in single-cell datasets.

"Highly variable genes (k = 2,244) were detected according to average expression (between 0.05 to 3) and dispersion (above 0.5) of the genes, followed by data scaling (subtracting the average expression) and centering (divided by standard deviation)."

More on FindVariableFeatures()- https://satijalab.org/seurat/reference/findvariablefeatures

```{r}
dim(livcan)

livcan <- FindVariableFeatures(
  livcan,
  selection.method = "mvp",
  mean.cutoff = c(0.05, 3),
  dispersion.cutoff = c(0.5, Inf),
  verbose = TRUE
)

dim(livcan)

# Identify the 10 most highly variable genes
top10_hvg <- head(VariableFeatures(livcan), 10)
# plot variable features with and without labels
plot1 <- VariableFeaturePlot(livcan)
plot2 <- LabelPoints(plot = plot1,
                     points = top10_hvg,
                     repel = TRUE)
plot1 + plot2
```
With the stringent cutoffs, only 1086 genes were classified as HVGs out of 5124, i.e., 4038 non-variable genes.

So, maybe they've used only the cutoffs mentioned in their methods. After using only those cutoffs, I got 2240 HVGs compared to their 2244 HVGs.

## Scaling the data

"...data scaling (subtracting the average expression) and centering (divided by standard deviation)"

Setting center to TRUE will center the expression for each feature by subtracting the average expression for that feature. Setting scale to TRUE will scale the expression level for each feature by dividing the centered feature expression levels by their standard deviations if center is TRUE and by their root mean square otherwise.

More on ScaleData()- https://satijalab.org/seurat/reference/scaledata

```{r}
all.genes <- rownames(livcan)

livcan <-
  ScaleData(
    livcan,
    do.center = TRUE,
    do.scale = TRUE,
    features = all.genes, 
    verbose = TRUE
  )
```

# Linear dimensional reduction

"Those variable genes were considered accounting for cell-to-cell differences, and were further used for PCA."

Now to perform PCA on the scaled data- By default, only the previously determined variable features are used as input, but can be defined using features argument if we wish to choose a different subset.

```{r}
length(VariableFeatures(object = livcan))
livcan <-
  RunPCA(livcan, features = VariableFeatures(object = livcan))
# Same can be achieved with livcan <- RunPCA(livcan) as it picks-up HVGs by default
# npcs, i.e.,Total Number of PCs to compute and store (50 by default)
```

Some different ways to visualize cells and features that define the PCA-

```{r}
# Way 1
print(livcan[["pca"]], dims = 1:5, nfeatures = 5)

# Way2
VizDimLoadings(livcan, dims = 1:2, reduction = "pca")

# Way3
DimPlot(livcan, reduction = "pca")

# Way4
DimHeatmap(livcan, dims = 1, cells = 500, balanced = TRUE)
DimHeatmap(livcan, dims = 1:20, cells = 500, balanced = TRUE)
```

# Determining the "dimensionality" of the dataset

‘Elbow plot’: a ranking of principle components based on the percentage of variance explained by each one (ElbowPlot() function). 

```{r}
ElbowPlot(livcan, ndims=20)
ElbowPlot(livcan, ndims=50)
```

- From the Elbow plot with 20 dims, we can observe an ‘elbow’ around PC19-20, suggesting that the majority of true signal is captured in the first 20 PCs.
- From the Elbow plot with 50 dims, we can observe an ‘elbow’ around PC44-45, suggesting that the majority of true signal is captured in the first 45 PCs.


An alternate approach would be to use JackStrawPlot().

```{r}
livcan <- JackStraw(livcan, num.replicate = 100, dims = 50, verbose = TRUE)
livcan <- ScoreJackStraw(livcan, dims = 1:50)

JackStrawPlot(livcan, dims = 1:50)
```

- From the JackStraw plot with 50 dims, it appears that there is a sharp drop-off in significance after the first 45 PCs (same can be said about a sharp drop-off about after 19 PCs or 26 PCs).

# Cluster the cells

Seurat uses a graph-based technique to cluster cells based on their gene expression profiles.

```{r}
livcan <- FindNeighbors(livcan, dims = 1:20)
livcan <- FindClusters(livcan, resolution = 0.325)

# Look at cluster IDs of the first 5 cells
head(Idents(livcan), 5)
```


# Non-linear dimensional reduction

The first 20 PCs were applied for t-SNE analysis according to the eigenvalues (data not shown).

The order is taken from Table 1 from <https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE125449> and PMID: 31588021:
  GSM4050085	S16_P10_LCP18  Patient no.  1  H18
  GSM4050086	S02_P01_LCP21  Patient no.  2  H21
  GSM4050087	S10_P05_LCP23  Patient no.  3  H23
  GSM4050088	S09_P04_LCP25  Patient no.  4  C25
  GSM4050089	S08_P03_LCP26  Patient no.  5  C26
  GSM4050090	S07_P02_LCP28  Patient no.  6  H28
  GSM4050091	S11_P06_LCP29  Patient no.  7  C29
  GSM4050092	S12_P07_LCP30  Patient no.  8  H30
  GSM4050093	S20_P12_LCP35  Patient no.  9  C35
  GSM4050094	S21_P13_LCP37  Patient no. 10  H37
  GSM4050095	S15_P09_LCP38  Patient no. 11  H38
  GSM4050096	S19_P11_LCP39  Patient no. 12  C39

```{r}
livcan <- RunTSNE(livcan, dims = 1:20)
DimPlot(
  livcan,
  reduction = "tsne",
  label = FALSE,
  group.by = "orig.ident",
  shuffle = TRUE,
  seed = 100,
  order = c(11, 9, 13, 12, 7, 6, 2, 3, 4, 5, 1, 10),
  # order = c(13, 12, 11, 10, 9, 7, 6, 5, 4, 3, 2, 1),
  # cols = DiscretePalette(12, palette = "alphabet2")
  cols = DiscretePalette(12, palette = "glasbey")
)
# DimPlot(livcan, reduction = "tsne")


# DimPlot(
#   livcan,
#   reduction = "tsne",
#   label = FALSE,
#   group.by = "ident",
#   split.by = "orig.ident",
#   ncol = 4,
#   # order = c(11, 9, 13, 12, 7, 6, 2, 3, 4, 5, 1, 10),
#   # combine = TRUE,
#   cols = DiscretePalette(12, palette = "alphabet2")
#   # cols = DiscretePalette(12)
# )
# # By default Seurat’s DimPlot() plots each group on top of the next which can make plots harder to interpret. DimPlot_scCustom sets shuffle = TRUE by default as I believe this setting is more often the visualization that provides the most clarity.
# # https://samuel-marsh.github.io/scCustomize/articles/Gene_Expression_Plotting.html
# DimPlot(seurat_object = marsh_human_pm, group.by = "sample_id")
# DimPlot_scCustom(seurat_object = livcan, group.by = "orig.ident")
```

```{r}
head(livcan@meta.data)
```


# R Markdown

For Markdown setup options, see <https://bookdown.org/yihui/rmarkdown/html-document.html>. 

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
summary(cars)
```

# Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.


# Session Information

```{r}
sessionInfo(package=NULL)
```

